crates/types/src/consensus/proposal.rs

```diff
@@ -80,6 +85,14 @@ impl Proposal {
 
         Bytes::from_iter(buf)
     }
+
+    pub fn flattened_pre_proposals(&self) -> Vec<PreProposal> {
+        self.preproposals
+            .iter()
+            .flat_map(|preproposal| preproposal.pre_proposals.clone())
+            .unique_by(|proposal| proposal.source)
```

Comment:
are we just taking any random proposal per source? or is there some other notion of uniqueness ordering we should be using

Reply:
no one per source. if someone double submits its a slashable event